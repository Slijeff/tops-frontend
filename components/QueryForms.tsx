import {
  Button,
  FormControl,
  FormHelperText,
  FormLabel,
  Input,
  Select,
  VStack,
} from "@chakra-ui/react";
import axios from "axios";
import React, { useEffect } from "react";
import { QueryFormProps } from "../types/QueryFrom";

type Intersection = {
  intersectionid: number;
  intersectionname: string;
};

const QueryForms = ({
    intersection,
    setIntersection,
    mode,
    setMode,
    start,
    setStart,
    end,
    setEnd,
    onQuery,
}: QueryFormProps) => {
  const [intersectionOptions, setIntersectionOptions] = React.useState<Intersection[]>();
  const getIntersection = async () => {
    await axios.get("/intersection/all").then((res) => {
      setIntersectionOptions(res.data.data);
    });
  };
  useEffect(() => {
    getIntersection();
  }, []);

  return (
    <VStack spacing={4}>
      <FormControl>
        <FormLabel>Select Intersection</FormLabel>
        <Select onChange={setIntersection}>
          {intersectionOptions?.map((i: Intersection) => {
            return (
              <option key={i.intersectionid} value={i.intersectionid}>
                {i.intersectionname}
              </option>
            );
          })}
        </Select>
      </FormControl>

      <FormControl>
        <FormLabel>Select Mode</FormLabel>
        <Select onChange={setMode}>
          <option value="history">Historical Mode</option>
          <option value="real">Real-time Mode</option>
        </Select>
      </FormControl>

      {mode === "history" && (
        <FormControl>
          <FormLabel>Start Date</FormLabel>
          <Input value={start} onChange={setStart} />
          <FormHelperText>Format: YYYY-MM-DD HH:MM:SS.SSS</FormHelperText>
        </FormControl>
      )}
      {mode === "history" && (
        <FormControl>
          <FormLabel>End Date</FormLabel>
          <Input value={end} onChange={setEnd} />
          {/* <FormHelperText>Format: YYYY-MM-DD HH:MM:SS.SSS</FormHelperText> */}
        </FormControl>
      )}

      {mode === "history" ? (
        <Button colorScheme={"teal"} w={"100%"} onClick={onQuery}>
          Query
        </Button>
      ) : (
        <Button colorScheme={"teal"} w={"100%"} isDisabled={true} isLoading={true} loadingText={"Auto Querying..."}/>
      )}
      
    </VStack>
  );
};
export default QueryForms;
