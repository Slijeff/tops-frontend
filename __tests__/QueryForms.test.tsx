import {describe, expect} from "@jest/globals";
import {fireEvent, render} from "@testing-library/react";
import QueryForms from "../components/QueryForms";
import * as axios from "axios";

jest.mock("axios")

describe("QueryForms rendering", () => {

    //@ts-ignore
    axios.get.mockResolvedValue({data: [{
        intersectionid: 1,
        intersectionname: "test"
      }]});

    test("should render QueryForms correctly", () => {
        
        const {getByText, container} = render(<QueryForms intersection={"int1"} setIntersection={(e) => {
        }} mode={"history"} setMode={(e) => {
        }} start={"2018-09-12 14:23:02"} setStart={(e) => {
        }} end={"2020-09-12 14:23:02"} setEnd={(e) => {
        }} onQuery={() => {}}/>)

        expect(container.firstChild).toMatchSnapshot()
        expect(getByText(/Historical Mode/i)).toBeTruthy()
        expect(getByText(/Real-time Mode/i)).toBeTruthy()
        expect(getByText(/Start Date/i)).toBeTruthy()
    });

    test("onQuery should fire accurately", () => {
        const handleClick = jest.fn()
        const {getByRole} = render(<QueryForms intersection={"int1"} setIntersection={(e) => {
        }} mode={"history"} setMode={(e) => {
        }} start={"2018-09-12 14:23:02"} setStart={(e) => {
        }} end={"2020-09-12 14:23:02"} setEnd={(e) => {
        }} onQuery={handleClick}/>)
        
        fireEvent.click(getByRole('button'))
        expect(handleClick).toBeCalledTimes(1)

    });
    
})