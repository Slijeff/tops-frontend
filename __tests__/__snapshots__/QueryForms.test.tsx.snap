// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryForms rendering should render QueryForms correctly 1`] = `
<div
  class="chakra-stack css-wulyol"
>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="field-:r0:"
      id="field-:r0:-label"
    >
      Select Intersection
    </label>
    <div
      class="chakra-select__wrapper css-42b2qy"
    >
      <select
        class="chakra-select css-1ik61og"
        id="field-:r0:"
      />
      <div
        class="chakra-select__icon-wrapper css-162mkon"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="field-:r1:"
      id="field-:r1:-label"
    >
      Select Mode
    </label>
    <div
      class="chakra-select__wrapper css-42b2qy"
    >
      <select
        class="chakra-select css-1ik61og"
        id="field-:r1:"
      >
        <option
          value="history"
        >
          Historical Mode
        </option>
        <option
          value="real"
        >
          Real-time Mode
        </option>
      </select>
      <div
        class="chakra-select__icon-wrapper css-162mkon"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="field-:r2:"
      id="field-:r2:-label"
    >
      Start Date
    </label>
    <input
      aria-describedby="field-:r2:-helptext"
      class="chakra-input css-0"
      id="field-:r2:"
      value="2018-09-12 14:23:02"
    />
    <div
      class="chakra-form__helper-text css-0"
      id="field-:r2:-helptext"
    >
      example: 2018-09-12 14:23:02
    </div>
  </div>
  <div
    class="chakra-form-control css-0"
    role="group"
  >
    <label
      class="chakra-form__label css-uvho5g"
      for="field-:r3:"
      id="field-:r3:-label"
    >
      End Date
    </label>
    <input
      aria-describedby="field-:r3:-helptext"
      class="chakra-input css-0"
      id="field-:r3:"
      value="2020-09-12 14:23:02"
    />
    <div
      class="chakra-form__helper-text css-0"
      id="field-:r3:-helptext"
    >
      example: 2018-09-12 14:23:02
    </div>
  </div>
  <button
    class="chakra-button css-uxt1e8"
    type="button"
  >
    Query
  </button>
</div>
`;
